ledger/stateproofverificationtracker.go

```diff
@@ -0,0 +1,244 @@
+
+func (spt *stateProofVerificationTracker) close() {
+	if spt.dbQueries.lookupStateProofVerificationData != nil {
+		spt.dbQueries.lookupStateProofVerificationData.Close()
+	}
+}
+
+func (spt *stateProofVerificationTracker) LookupVerificationData(stateProofLastAttestedRound basics.Round) (*ledgercore.StateProofVerificationData, error) {
+	spt.stateProofVerificationMu.RLock()
+	defer spt.stateProofVerificationMu.RUnlock()
+
+	if len(spt.trackedCommitData) == 0 || stateProofLastAttestedRound < spt.trackedCommitData[0].verificationData.TargetStateProofRound {
+		return spt.dbQueries.lookupData(stateProofLastAttestedRound)
+	}
+
+	if stateProofLastAttestedRound <= spt.trackedCommitData[len(spt.trackedCommitData)-1].verificationData.TargetStateProofRound {
+		verificationData, err := spt.lookupDataInTrackedMemory(stateProofLastAttestedRound)
+		return &verificationData, err
+	}
+
+	return &ledgercore.StateProofVerificationData{}, errStateProofVerificationDataNotYetGenerated
+}
+
+func (spt *stateProofVerificationTracker) committedRoundToLatestCommitDataIndex(committedRound basics.Round) int {
+	latestCommittedDataIndex := -1
+
+	for index, data := range spt.trackedCommitData {
+		if data.confirmedRound <= committedRound {
+			latestCommittedDataIndex = index
+		}
+	}
+
+	return latestCommittedDataIndex
+}
+
+func (spt *stateProofVerificationTracker) committedRoundToLatestDeleteDataIndex(committedRound basics.Round) int {
+	latestCommittedDataIndex := -1
+
+	for index, data := range spt.trackedDeleteData {
+		if data.confirmedRound <= committedRound {
+			latestCommittedDataIndex = index
+		}
+	}
+
+	return latestCommittedDataIndex
+}
+
+func (spt *stateProofVerificationTracker) lookupDataInTrackedMemory(stateProofLastAttestedRound basics.Round) (ledgercore.StateProofVerificationData, error) {
+	for _, commitData := range spt.trackedCommitData {
+		if commitData.verificationData.TargetStateProofRound == stateProofLastAttestedRound {
+			return commitData.verificationData, nil
+		}
+	}
+
+	return ledgercore.StateProofVerificationData{}, errStateProofVerificationDataNotFound
+}
+
+func (spt *stateProofVerificationTracker) insertCommitData(blk *bookkeeping.Block) {
+	verificationData := ledgercore.StateProofVerificationData{
+		VotersCommitment:      blk.StateProofTracking[protocol.StateProofBasic].StateProofVotersCommitment,
+		ProvenWeight:          blk.StateProofTracking[protocol.StateProofBasic].StateProofOnlineTotalWeight,
+		TargetStateProofRound: blk.Round() + basics.Round(blk.ConsensusProtocol().StateProofInterval),
+	}
+
+	commitData := verificationCommitData{
+		confirmedRound:   blk.Round(),
+		verificationData: verificationData,
+	}
+	spt.trackedCommitData = append(spt.trackedCommitData, commitData)
+}
+
+func (spt *stateProofVerificationTracker) insertDeleteData(blk *bookkeeping.Block, delta *ledgercore.StateDelta) {
+	deletionData := verificationDeleteData{
+		confirmedRound:      blk.Round(),
+		stateProofNextRound: delta.StateProofNext,
```

Comment:
I disagree - in my view, the input we receive upon deletion represents the earliest state proof round we're supposed to keep tracking. I've renamed some variables in an attempt to make things clearer, and wouldn't mind adding some comments explaining it further.
I think that subtracting the interval would not increase accuracy and could be problematic in case of interval changes.